src/Infrastructure/Process/ExecutableFinder.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace PhpTuf\ComposerStager\Infrastructure\Process;
+
+use PhpTuf\ComposerStager\Exception\IOException;
+use Symfony\Component\Process\ExecutableFinder as SymfonyExecutableFinder;
+
+/**
+ * @internal
+ */
+class ExecutableFinder
+{
+    /**
+     * @var \PhpTuf\ComposerStager\Exception\IOException[]|string[]
+     */
+    private $cache = [];
+
+    /**
+     * @var \Symfony\Component\Process\ExecutableFinder
+     */
+    private $symfonyExecutableFinder;
+
+    public function __construct(SymfonyExecutableFinder $symfonyExecutableFinder)
+    {
+        $this->symfonyExecutableFinder = $symfonyExecutableFinder;
+    }
+
+    /**
+     * @throws \PhpTuf\ComposerStager\Exception\IOException
+     */
+    public function find(string $name): string
+    {
+        $cache = $this->getCache($name);
```

Comment:
It might be, except that it also caches exceptions. I've toyed with a few other designs to improve readability, but I'm afraid that all the code added by any obvious alternatives actually obscures the intent. If you think there's a better pattern or idiom I'm open to hearing it.

Reply:
We discussed this at scrum: we're fine with it for now but may come back and refine the solution later.